ehsm_kms_service/secret_manager_apis.js

```diff
@@ -265,9 +360,709 @@ const createSecret = async (res, appid, payload, DB) => {
+        // if not use versionStage = SM_SECRET_VERSION_STAGE_CURRENT
+        if (versionId) {
+            selector = {
+                appid: appid,
+                secretName: secretName,
+                versionId: versionId,
+                deletedFlag: false
+            }
+        } else {
+            selector = {
+                appid: appid,
+                secretName: secretName,
+                versionStage: SM_SECRET_VERSION_STAGE_CURRENT,
+                deletedFlag: false
+            }
+        }
+
+        // query secret_version_data
+        const query = {
+            selector,
+            fields: ['secretData', 'versionId', 'createTime'],
+            limit: 1,
+        }
+        let query_result = await DB.partitionedFind('secret_version_data', query)
+
+        // build result
+        let result = {}
+        if (query_result.docs.length > 0) {
+            let secretData = query_result.docs[0]['secretData']
+            let createTime = query_result.docs[0]['createTime']
+            versionId = query_result.docs[0]['versionId']
+
+            //query encryptionKeyId
+            const query_encryptionKeyId = {
+                selector: {
+                    appid,
+                    secretName,
+                },
+                fields: ['encryptionKeyId'],
+                limit: 1,
+            }
+            const secret_metadata_res = await DB.partitionedFind('secret_metadata', query_encryptionKeyId)
+            if (secret_metadata_res.docs.length == 0) {
+                encryptionKeyId = ""
+            } else {
+                encryptionKeyId = secret_metadata_res.docs[0].encryptionKeyId
+            }
+
+            // query SM_Masterkey
+            let sm_masterKey
+            let retCMK = await get_SM_Masterkey(DB, appid, encryptionKeyId)
+            if (retCMK.isFind) {
+                sm_masterKey = retCMK.sm_masterKey
+            } else {
+                res.send(_result(400, retCMK.msg))
+                return
+            }
+
+            // decrypt secretData
+            const secretData_decypt_result = napi_result(cryptographic_apis.Decrypt, res, [sm_masterKey, secretData, ''])
+            if (secretData_decypt_result) {
+                secretData = secretData_decypt_result.result['plaintext']
+            } else {
+                res.send(_result(400, 'decypt secretData error', {}))
+                return
+            }
+
+            result['secretName'] = base64_decode(secretName)
+            result['secretData'] = base64_decode(secretData)
+            result['versionId'] = versionId
+            result['createTime'] = createTime
+        }
+        res.send(_result(200, 'successful', result))
+        return
+    } catch (e) {
+        console.info('getSecretValue :: ', e)
+        res.send(_result(500, 'Server internal error, please contact the administrator.'))
+        return
+    }
+}
+
+/**
+ * Restores a deleted secret.
```

Comment:
we do not check whether the secret data is truly deleted, right?

Reply:
yes